File: daily_problems/2026/09/0902/personal_submission/cf105223g_yawn_sean.py
# Submission link: https://codeforces.com/gym/105223/submission/389067167
def main():
    n = II()
    nums = LII()
    
    cnt = [0] * 30
    
    for x in nums:
        for i in range(30):
            cnt[i] += x >> i & 1
    
    mod = 10 ** 9 + 7
    pw3 = [1] * (n + 1)
    
    for i in range(n):
        pw3[i + 1] = pw3[i] * 3 % mod
    
    rev2 = (mod + 1) // 2
    
    q = II()
    outs = []
    
    for _ in range(q):
        idx, val = MII()
        idx -= 1
        
        for i in range(30):
            cnt[i] -= nums[idx] >> i & 1
    
        nums[idx] = val
        
        for i in range(30):
            cnt[i] += nums[idx] >> i & 1
        
        ans = 0
        
        for i in range(29, -1, -1):
            ans = (ans * 2 + (pw3[n] - pw3[n - cnt[i]])) % mod
        
        outs.append(ans * rev2 % mod)
    
    print('\n'.join(map(str, outs)))

File: daily_problems/2026/09/0902/personal_submission/cf105223g_yawn_sean.cpp
#include <bits/stdc++.h>
// #include "atcoder/all"
// #pragma GCC optimize("Ofast")
// #pragma GCC optimize("unroll-loops")
// #pragma GCC target("sse,sse2,sse3,ssse3,sse4,popcnt,abm,mmx,avx2,tune=native")

using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n;
	cin >> n;

	vector<int> nums(n), cnt(30, 0);

	for (auto &x: nums) {
		cin >> x;
		for (int i = 0; i < 30; i ++) {
			cnt[i] += x >> i & 1;
		}
	}

	int mod = 1e9 + 7, rev2 = (mod + 1) / 2;
	vector<int> pw3(n + 1, 1);
	for (int i = 0; i < n; i ++) pw3[i + 1] = 3ll * pw3[i] % mod;

	int q;
	cin >> q;

	while (q --) {
		int idx, val;
		cin >> idx >> val;
		idx --;

		for (int i = 0; i < 30; i ++) cnt[i] -= nums[idx] >> i & 1;
		nums[idx] = val;
		for (int i = 0; i < 30; i ++) cnt[i] += nums[idx] >> i & 1;

		int ans = 0;
		for (int i = 29; i >= 0; i --) {
			ans = 2 * ans % mod;
			ans = (ans + pw3[n]) % mod;
			ans = (ans + mod - pw3[n - cnt[i]]) % mod;
		}

		cout << 1ll * ans * rev2 % mod << '\n';
	}

	return 0;
}